src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -203,8 +203,9 @@ trait TypeAssigner {
    */
   def selectionType(site: Type, name: Name, pos: Position)(implicit ctx: Context): Type = {
     val mbr = site.member(name)
-    if (reallyExists(mbr)) site.select(name, mbr)
-    else if (site.derivesFrom(defn.DynamicClass) && !Dynamic.isDynamicMethod(name)) {
+    lazy val canBeDynamicMethod = site.derivesFrom(defn.DynamicClass) && !Dynamic.isDynamicMethod(name)
+    if (reallyExists(mbr) && (mbr.accessibleFrom(site).exists || !canBeDynamicMethod)) site.select(name, mbr)
```

Comment:
The issue here is that `site.member(name)` can return a private member that is not accessible from this type (for example a private def in one of the parents). These pass the the checks in `selectionType` and fail later when the accessibility of the member is checked. The condition above is there to check that the selected member is in fact accessible. In case it is not and that this is a Dynamic we need to try one of the dynamics transformation.

See run/t5040.scala